src/lib/Inline.hs

```diff
@@ -0,0 +1,497 @@
+    --     for k. ys.k
+    -- really does have two available indexing levels, because the ys binding is
+    -- inlinable into the result of the outer `for`; but the below function will
+    -- not detect it, and thus we will decline to inline `xs` into a context
+    -- where two levels of indexing are required.  However, we can recover the
+    -- desired effect by running a second pass of occurrence analysis and
+    -- inlining.
+    --
+    -- A more subtle case occurs with
+    --   xs = for i.
+    --     ys = for j. <something>
+    --     view k. ys.k
+    -- (note the `view`).  In this case, if we went ahead and inlined `xs` into
+    -- a context that required two levels of indexing, we would temporarily
+    -- duplicate work, but a second pass of occurence analysis and inlining
+    -- would fix it by inlining `ys`.  However, if we decline to inline here,
+    -- running inlining on the body of `for i.` will not inline into the `view`,
+    -- because in general views are supposed to be duplicable all over.  Maybe
+    -- the solution is to just not have view atoms in the IR by this point,
+    -- since their main purpose is to force inlining in the simplifier, and if
+    -- one just stuck like this it has become equivalent to a `for` anyway.
+    ixDepthExpr :: Expr r n -> Int
+    ixDepthExpr (Hof (For _ _ (Lam (LamExpr _ body)))) = 1 + ixDepthBlock body
+    ixDepthExpr (Atom (TabLam (TabLamExpr _ body))) = 1 + ixDepthBlock body
+    ixDepthExpr _ = 0
+    ixDepthBlock :: Block r n -> Int
+    ixDepthBlock (exprBlock -> (Just expr)) = ixDepthExpr expr
+    ixDepthBlock (AtomicBlock result) = ixDepthExpr $ Atom result
+    ixDepthBlock _ = 0
+
+-- Should we decide to inline this binding wherever it appears, before we even
+-- know the expression?  "Yes" only if we know it only occurs once, and in a
+-- context where inlining it does not duplicate work.
+preInlineUnconditionally :: LetAnn -> Bool
+preInlineUnconditionally = \case
+  PlainLet -> False  -- "Missing occurrence annotation"
+  NoInlineLet -> False
+  OccInfoPure (UsageInfo s (0, d)) | s <= One && d <= One -> True
+  OccInfoPure _ -> False
+  OccInfoImpure _ -> False
+
+-- A context in which an E-kinded thing is to be reconstructed.  This amounts to
+-- a defunctionalization of the interesting part of rebuilding an expression,
+-- which supports now-local optimizations in `reconstruct`.  Read `reconstruct`
+-- together with this data type: `reconstruct` is an interpreter for it.
+--
+-- Note: This pattern of inserting an object into a context can do local
+-- optimizations that would otherwise be hidden from a peephole optimizer by
+-- intervening bindings.  For example, table indexing only permits an Atom in
+-- the array position, but `reconstruct` can check whether it's trying to insert
+-- a `for` expression into that spot and perform the beta reduction immediately,
+-- instead of emitting the binding.
+data Context (from::E) (to::E) (o::S) where
```

Comment:
This is a really clever trick! I wouldn't have thought of doing it that way.

Reply:
Thank you!  It did take some time to work out.